packages/edit-navigation/src/components/menu-locations-editor/use-menu-locations.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * External dependencies
+ */
+import { map, flatMap, forEach, filter, groupBy } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { useDispatch } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
+
+export default function useMenuLocations() {
+	const { saveMenu } = useDispatch( 'core' );
+	const [ menuLocations, setMenuLocations ] = useState( null );
+
+	// a local state which maps menus to locations
+	// so that we can send one call per menu when
+	// updating locations, otherwise, without this local state
+	// we'd send one call per location
+	const [ menuLocationMap, setMenuLocationMap ] = useState( {} );
+
+	const fetchmenuLocations = async () => {
+		const path = '/__experimental/menu-locations';
+		const apiLocations = await apiFetch( {
+			path,
+			method: 'GET',
+		} );
+		return flatMap( apiLocations );
+	};
+
+	const initmenuLocations = async () => {
```

Comment:
nit: it seems like we don't need `fetchmenuLocations` and `initmenuLocations` to be two separate functions? Also `initmenuLocations` -> `initMenuLocations` (camelCase)